Add documentation about temporary compile outputs (#23251)
gradle/gradle: subprojects/docs/src/docs/userguide/jvm/java_plugin.adoc

* Having a source structure that does not match the package names, while legal for compilation, might end up causing trouble in the toolchain.
    Even more if annotation processing and <<build_cache.adoc#build_cache,caching>> are involved.

> Additionally, Gradle might temporary change the output location while running incremental compilation. This might affect some annotation processors that inspect output locations and which allow users to wire some action to file paths (e.g. `-XepExcludedPaths` in Error Prone). You can disable this Gradle behaviour by setting link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task. Alternatively you can try specify temporary output path to annotation processor. Output location mapping to temporary output locations is listed in the next table:

Reviewer: ❌ Can you do one line per sentence?

Author: Should be fixed now

> Additionally, Gradle might temporary change the output location while running incremental compilation. This might affect some annotation processors that inspect output locations and which allow users to wire some action to file paths (e.g. `-XepExcludedPaths` in Error Prone). You can disable this Gradle behaviour by setting link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task. Alternatively you can try specify temporary output path to annotation processor. Output location mapping to temporary output locations is listed in the next table:

Reviewer: 💄  
```suggestion
Additionally, Gradle might temporary change the output location while running incremental compilation. This might affect some annotation processors that inspect output locations and which allow users to wire some action to file paths (e.g. `-XepExcludedPaths` in Error Prone). You can disable this Gradle behaviour by setting link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task. Alternatively, you can try to specify a temporary output path to the annotation processor. The next table shows the mapping from output location to temporary output location:
```

> .Compiler output to temporary output locations mapping

Reviewer: 💄 
```suggestion
.Compiler output to temporary output location
```

|===

> | JavaCompile location property | Example output location | Temporary output location

Reviewer: ```suggestion
| JavaCompile location property | Default output location | Temporary output location
```

💄 I wonder if "Default" is better here.

* Having a source structure that does not match the package names, while legal for compilation, might end up causing trouble in the toolchain.
    Even more if annotation processing and <<build_cache.adoc#build_cache,caching>> are involved.

> Additionally, Gradle might temporary change the output location while running incremental compilation.

Reviewer 2: ```suggestion
Additionally, Gradle might temporarily change the output location while running incremental compilation.
```

> This might affect some annotation processors that inspect output locations and which allow users to wire some action to file paths (e.g. `-XepExcludedPaths` in Error Prone).

Reviewer 2: ```suggestion
This might affect some annotation processors that inspect output locations or accept file paths as arguments (e.g. `-XepExcludedPaths` in Error Prone).
```

> You can disable this Gradle behaviour by setting link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task.

Reviewer 2: ```suggestion
You can disable this behaviour by setting link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task.
```

> Additionally, Gradle might temporarily change the output location while running incremental compilation.
> This might affect some annotation processors that inspect output locations or accept file paths as arguments (e.g. `-XepExcludedPaths` in Error Prone).
> You can disable this behaviour by setting link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task.
> Alternatively, you can try to specify a temporary output path to the annotation processor.

Reviewer 3: ```suggestion
Additionally, Gradle might temporarily change the output location while running incremental compilation. 
This might affect some annotation processors that inspect output locations or accept file paths as arguments (e.g., `-XepExcludedPaths` in Error Prone).
There are two options:
 - You can disable `incremental after failure` by setting the link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task.
-  You can keep `incremental after failure` working by passing a temporary output value to the annotation processor as a parameter so the annotation processor is aware of that path. In case of Error Prone, pass a value to `-XepExcludedPaths`. For example:
Before, you might have had: `-XepExcludedPaths=.*/build/generated/.*"`. With the new behavior, they might need to do
`-XepExcludedPaths=.*/(build/generated|compileJava/compileTransaction/annotation-output)/.*"`.

Note the need for this argument is planned to be resolved in [#23226](https://github.com/gradle/gradle/pull/23226).
```

Reviewer 3: not in love with it, feel free to change

Author: Looks good to me. I would just exclude the PR, since it is not yet complete and we might find some surprises.

Author: I removed the note and applied a suggestion

> Alternatively, you can try to specify a temporary output path to the annotation processor.

Reviewer 2: I'm not sure I understand this sentence properly. Is there some mechanism for users to configure what the temporary output path is? Or is this a more general comment for annotation processors which themselves can be configured with custom file paths?

Author: Thanks for the review, I believe this might not be clear to a lot of users.

Users can't change that temporary output, but they can make annotation aware of it.
So basically what I want to say here is: 
The user has two options:
1.) They can just disable the behavior, done.
2.) Or they pass this value to the annotation processor as a parameter, so the annotation processor is aware of that path. In case of Error Prone they can pass that value to `-XepExcludedPaths`. Example:
Before they had: `-XepExcludedPaths=.*/build/generated/.*"`. With new behaviour that might not be good, so they should add `-XepExcludedPaths=.*/(build/generated|<task-name>/compileTransaction/annotation-output)/.*"` (tbh, l hope in Gradle 8.1 we get rid of that workaround with https://github.com/gradle/gradle/pull/23226).

Reviewer 3: This extra explanation is nice, are you adding more into the doc for the users?

Reviewer 2: Perhaps: 

```
This might affect some annotation processors that inspect output locations or
accept file paths as arguments. For example, Error Prone's `-XepExcludedPaths` 
parameter can be updated to `.*/(build/generated|compileJava/compileTransaction/annotation-output)/.*` 
in order to make the processor aware of the new temporary directory. 
```

> The next table shows the mapping from output location to temporary output location:

Reviewer 2: ```suggestion
The following table shows the mapping from default output location to temporary output location:
```

* Having a source structure that does not match the package names, while legal for compilation, might end up causing trouble in the toolchain.
    Even more if annotation processing and <<build_cache.adoc#build_cache,caching>> are involved.

Additionally, Gradle might temporarily change the output location while running incremental compilation.
This might affect some annotation processors that inspect output locations or accept file paths as arguments (e.g., `-XepExcludedPaths` in Error Prone).
There are two options:

>  - You can disable `incremental after failure` by setting the link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task.

Reviewer 2: ```suggestion
 - Disable `incremental after failure` by setting link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html#getIncrementalAfterFailure--[`options.incrementalAfterFailure`] to `false` on the JavaCompile task.
```

> -  You can keep `incremental after failure` working by passing a temporary output value to the annotation processor as a parameter so the annotation processor is aware of that path. In case of Error Prone, pass a value to `-XepExcludedPaths`. For example:
> Before, you might have had: `-XepExcludedPaths=.\*/build/generated/.*`.
> With the new behavior, they might need to do +
> `-XepExcludedPaths=.\*/(build/generated|compileJava/compileTransaction/annotation-output)/.*`.

Reviewer 2: ```suggestion
-  Keep `incremental after failure` enabled by instead passing a temporary output value to the annotation processor as a parameter so that the annotation processor is aware of the temporary path. 

In the case of Error Prone, `-XepExcludedPaths` must be set. Given an existing value of `-XepExcludedPaths=.\*/build/generated/.*`, the updated value would be
`-XepExcludedPaths=.\*/(build/generated|compileJava/compileTransaction/annotation-output)/.*`.
```

Author: Sounds good to me
